File: packages/compiler-rs/src/utils/walk.rs
use oxc_allocator::{CloneIn, TakeIn};
use oxc_ast::ast::{
  ArrowFunctionExpression, AssignmentTargetMaybeDefault, AssignmentTargetProperty,
  BindingIdentifier, BindingPattern, BindingPatternKind, BlockStatement, CatchClause, Expression,
  ForInStatement, ForOfStatement, ForStatement, ForStatementInit, ForStatementLeft, Function,
  FunctionBody, Program, Statement, VariableDeclarationKind,
};
use oxc_semantic::SemanticBuilder;
use oxc_traverse::{Ancestor, Traverse, TraverseAncestry, TraverseCtx, traverse_mut};
use std::collections::{HashMap, HashSet};

use napi::bindgen_prelude::Either3;
use oxc_ast::{AstKind, ast::IdentifierReference};
use oxc_span::{GetSpan, SPAN, Span};

use crate::{
  generate::CodegenContext, transform::TransformContext, utils::check::is_referenced_identifier,
};

/**
 * Modified from https://github.com/vuejs/core/blob/main/packages/compiler-core/src/babelUtils.ts
 * To support browser environments and JSX.
 *
 * https://github.com/vuejs/core/blob/main/LICENSE
 *
 * Return value indicates whether the AST walked can be a constant
 */
pub struct WalkIdentifiers<'a, 'ctx> {
  known_ids: HashMap<String, u32>,
  include_all: bool,
  context: &'a CodegenContext<'ctx>,
  on_identifier: Box<
    dyn FnMut(
        &mut IdentifierReference<'a>,
        &Ancestor,
        &TraverseAncestry<'a>,
        bool,
        bool,
      ) -> Option<Expression<'a>>
      + 'a,
  >,
  scope_ids_map: HashMap<Span, HashSet<String>>,
  roots: Vec<*mut Expression<'a>>,
}

impl<'a, 'ctx> WalkIdentifiers<'a, 'ctx> {
  pub fn new(
    context: &'a CodegenContext<'ctx>,
    on_identifier: Box<
      dyn FnMut(
          &mut IdentifierReference<'a>,
          &Ancestor,
          &TraverseAncestry<'a>,
          bool,
          bool,
        ) -> Option<Expression<'a>>
        + 'a,
    >,
    include_all: bool,
  ) -> Self {
    Self {
      context,
      on_identifier,
      include_all,
      known_ids: HashMap::new(),
      scope_ids_map: HashMap::new(),
      roots: vec![],
    }
  }

  pub fn traverse(&mut self, expression: Expression<'a>) -> Expression<'a> {
    let ast = &self.context.ast;
    let source_text = expression.span().source_text(self.context.ir.source);
    let program = &mut ast.program(
      SPAN,
      self.context.options.source_type,
      source_text,
      ast.vec(),
      None,
      ast.vec(),
      ast.vec_from_array([ast.statement_expression(SPAN, expression)]),
    );
    traverse_mut(
      self,
      ast.allocator,
      program,
      SemanticBuilder::new()
        .build(program)
        .semantic
        .into_scoping(),
      (),
    );
    let Statement::ExpressionStatement(stmt) = &mut program.body[0] else {
      unreachable!();
    };
    if self.roots.is_empty() {
      stmt.expression.take_in(ast.allocator)
    } else {
      stmt.expression.clone_in(ast.allocator)
    }
  }

  fn exit_node(&mut self, span: &Span, ctx: &mut oxc_traverse::TraverseCtx<'a, ()>) {
    let known_ids = &mut self.known_ids;
    if !matches!(ctx.parent(), Ancestor::None)
      && let Some(scope_ids) = self.scope_ids_map.get(span)
    {
      for id in scope_ids {
        if let Some(size) = known_ids.get(id) {
          known_ids.insert(id.clone(), size - 1);
          if known_ids[id] == 0 {
            known_ids.remove(id);
          }
        }
      }
    }
  }

  fn on_identifier_reference(
    &mut self,
    id: &mut IdentifierReference<'a>,
    ctx: &mut TraverseCtx<'a, ()>,
  ) -> Option<Expression<'a>> {
    if id.span.eq(&SPAN) {
      return None;
    }
    let is_local = self.known_ids.contains_key(id.name.as_str());
    let is_refed = is_referenced_identifier(id, &ctx.ancestry);
    if self.include_all || (is_refed && !is_local) {
      self.on_identifier.as_mut()(id, &ctx.parent(), &ctx.ancestry, is_refed, is_local)
    } else {
      None
    }
  }
}

impl<'a> Traverse<'a, ()> for WalkIdentifiers<'a, '_> {
  fn enter_assignment_target_property(
    &mut self,
    node: &mut AssignmentTargetProperty<'a>,
    ctx: &mut TraverseCtx<'a, ()>,
  ) {
    let ast = self.context.ast;
    match node {
      // ;({ baz } = bar)
      AssignmentTargetProperty::AssignmentTargetPropertyIdentifier(id) => {
        if let Some(replacer) = self.on_identifier_reference(&mut id.binding, ctx) {
          *node = ast.assignment_target_property_assignment_target_property_property(
            SPAN,
            ast.property_key_static_identifier(id.binding.span, id.binding.name),
            match replacer {
              Expression::Identifier(replacer) => {
                AssignmentTargetMaybeDefault::AssignmentTargetIdentifier(replacer)
              }
              Expression::StaticMemberExpression(replacer) => {
                AssignmentTargetMaybeDefault::StaticMemberExpression(replacer)
              }
              _ => unimplemented!(),
            },
            false,
          );
        };
      }
      // ;({ baz: baz } = bar)
      AssignmentTargetProperty::AssignmentTargetPropertyProperty(property) => {
        match &mut property.binding {
          AssignmentTargetMaybeDefault::AssignmentTargetIdentifier(id) => {
            if let Some(replacer) = self.on_identifier_reference(id, ctx) {
              property.binding = match replacer {
                Expression::Identifier(replacer) => {
                  AssignmentTargetMaybeDefault::AssignmentTargetIdentifier(replacer)
                }
                Expression::StaticMemberExpression(replacer) => {
                  AssignmentTargetMaybeDefault::StaticMemberExpression(replacer)
                }
                _ => unimplemented!(),
              };
            }
          }
          _ => unreachable!(),
        };
      }
    }
  }
  fn enter_expression(&mut self, node: &mut Expression<'a>, ctx: &mut TraverseCtx<'a, ()>) {
    if let Expression::Identifier(id) = node {
      if let Some(replacer) = self.on_identifier_reference(id, ctx) {
        *node = replacer;
      }
    } else if matches!(node, Expression::JSXElement(_) | Expression::JSXFragment(_)) {
      if self.context.options.interop {
        for node in ctx.ancestors() {
          if let Ancestor::CallExpressionArguments(node) = node {
            let name = node.callee().span().source_text(self.context.ir.source);
            if name == "defineVaporComponent" {
              break;
            } else if name == "defineComponent" {
              return;
            }
          }
        }
      }
      self.roots.push(node as *mut Expression);
    }
  }
  fn exit_program(&mut self, _: &mut Program<'a>, _: &mut TraverseCtx<'a, ()>) {
    let allocator = self.context.ast.allocator;
    let transform_context = self.context.transform_cotext;
    unsafe {
      for root in &mut self.roots {
        let root = &mut **root;
        let context: *mut TransformContext =
          &mut TransformContext::new(allocator, self.context.options);
        *(&mut *context).in_v_once.borrow_mut() = *transform_context.in_v_once.borrow();
        *(&mut *context).in_v_for.borrow_mut() = *transform_context.in_v_for.borrow();
        let source = &self.context.ir.source[..root.span().end as usize];
        *root = (&*context).transform(root.take_in(allocator), source);
      }
    }
  }
  fn exit_identifier_reference(
    &mut self,
    node: &mut IdentifierReference<'a>,
    ctx: &mut TraverseCtx<'a, ()>,
  ) {
    self.exit_node(&node.span, ctx)
  }

  fn enter_function(&mut self, node: &mut Function<'a>, _: &mut TraverseCtx<'a, ()>) {
    if let Some(scope_ids) = self.scope_ids_map.get(&node.span) {
      for id in scope_ids {
        mark_known_ids(id.clone(), &mut self.known_ids);
      }
    } else {
      // walk function expressions and add its arguments to known identifiers
      // so that we don't prefix them
      for p in &node.params.items {
        for id in extract_identifiers(&p.pattern, Vec::new()) {
          mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map)
        }
      }
    }
  }
  fn exit_function(&mut self, node: &mut Function<'a>, ctx: &mut TraverseCtx<'a, ()>) {
    self.exit_node(&node.span, ctx)
  }

  fn enter_arrow_function_expression(
    &mut self,
    node: &mut ArrowFunctionExpression<'a>,
    _: &mut TraverseCtx<'a, ()>,
  ) {
    if let Some(scope_ids) = self.scope_ids_map.get(&node.span) {
      for id in scope_ids {
        mark_known_ids(id.clone(), &mut self.known_ids);
      }
    } else {
      // walk function expressions and add its arguments to known identifiers
      // so that we don't prefix them
      for p in &node.params.items {
        for id in extract_identifiers(&p.pattern, Vec::new()) {
          mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map)
        }
      }
    }
  }
  fn exit_arrow_function_expression(
    &mut self,
    node: &mut ArrowFunctionExpression<'a>,
    ctx: &mut TraverseCtx<'a, ()>,
  ) {
    self.exit_node(&node.span, ctx)
  }

  fn enter_function_body(&mut self, node: &mut FunctionBody<'a>, _: &mut TraverseCtx<'a, ()>) {
    if let Some(scope_ids) = self.scope_ids_map.get(&node.span) {
      for id in scope_ids {
        mark_known_ids(id.clone(), &mut self.known_ids);
      }
    } else {
      walk_block_declarations(&node.statements, |id| {
        mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map);
      });
    }
  }
  fn exit_function_body(&mut self, node: &mut FunctionBody<'a>, ctx: &mut TraverseCtx<'a, ()>) {
    self.exit_node(&node.span, ctx);
  }

  fn enter_block_statement(&mut self, node: &mut BlockStatement<'a>, _: &mut TraverseCtx<'a, ()>) {
    if let Some(scope_ids) = self.scope_ids_map.get(&node.span) {
      for id in scope_ids {
        mark_known_ids(id.clone(), &mut self.known_ids);
      }
    } else {
      // #3445 record block-level local variables
      walk_block_declarations(&node.body, |id| {
        mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map);
      });
    }
  }
  fn exit_block_statement(&mut self, node: &mut BlockStatement<'a>, ctx: &mut TraverseCtx<'a, ()>) {
    self.exit_node(&node.span, ctx);
  }

  fn enter_catch_clause(&mut self, node: &mut CatchClause<'a>, _: &mut TraverseCtx<'a, ()>) {
    if let Some(param) = &node.param {
      for id in extract_identifiers(&param.pattern, vec![]) {
        mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map);
      }
    }
  }
  fn exit_catch_clause(&mut self, node: &mut CatchClause<'a>, ctx: &mut TraverseCtx<'a, ()>) {
    self.exit_node(&node.span, ctx);
  }

  fn enter_for_statement(&mut self, node: &mut ForStatement<'a>, _: &mut TraverseCtx<'a, ()>) {
    walk_for_statement(Either3::A(&node), true, &mut |id| {
      mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map);
    })
  }
  fn exit_for_statement(&mut self, node: &mut ForStatement<'a>, ctx: &mut TraverseCtx<'a, ()>) {
    self.exit_node(&node.span, ctx);
  }
  fn enter_for_in_statement(&mut self, node: &mut ForInStatement<'a>, _: &mut TraverseCtx<'a, ()>) {
    walk_for_statement(Either3::B(&node), true, &mut |id| {
      mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map);
    })
  }
  fn exit_for_in_statement(
    &mut self,
    node: &mut ForInStatement<'a>,
    ctx: &mut TraverseCtx<'a, ()>,
  ) {
    self.exit_node(&node.span, ctx);
  }

  fn enter_for_of_statement(&mut self, node: &mut ForOfStatement<'a>, _: &mut TraverseCtx<'a, ()>) {
    walk_for_statement(Either3::C(&node), true, &mut |id| {
      mark_scope_identifier(node.span, id, &mut self.known_ids, &mut self.scope_ids_map);
    })
  }
  fn exit_for_of_statement(
    &mut self,
    node: &mut ForOfStatement<'a>,
    ctx: &mut TraverseCtx<'a, ()>,
  ) {
    self.exit_node(&node.span, ctx);
  }
}

pub fn mark_known_ids(name: String, known_ids: &mut HashMap<String, u32>) {
  if let Some(ids) = known_ids.get(&name) {
    known_ids.insert(name, ids + 1);
  } else {
    known_ids.insert(name, 1);
  }
}

pub fn mark_scope_identifier<'a>(
  node_span: Span,
  child: &BindingIdentifier,
  known_ids: &mut HashMap<String, u32>,
  scope_ids_map: &mut HashMap<Span, HashSet<String>>,
) {
  let name = child.name.to_string();
  if let Some(scope_ids) = scope_ids_map.get_mut(&node_span) {
    if scope_ids.contains(&name) {
      return;
    } else {
      scope_ids.insert(name.clone());
    }
  } else {
    scope_ids_map.insert(node_span, HashSet::from([name.clone()]));
  }
  mark_known_ids(name, known_ids);
}

pub fn walk_function_params<'a>(
  node: &'a AstKind,
  mut on_ident: impl FnMut(&'a BindingIdentifier) + 'a,
) {
  let params = match node {
    AstKind::Function(node) => &node.params.items,
    AstKind::ArrowFunctionExpression(node) => &node.params.items,
    _ => panic!(""),
  };
  for p in params {
    for id in extract_identifiers(&p.pattern, Vec::new()) {
      on_ident(id)
    }
  }
}

pub fn extract_identifiers<'a>(
  node: &'a BindingPattern<'a>,
  mut identifiers: Vec<&'a BindingIdentifier<'a>>,
) -> Vec<&'a BindingIdentifier<'a>> {
  match &node.kind {
    BindingPatternKind::BindingIdentifier(node) => identifiers.push(node.as_ref()),
    BindingPatternKind::ObjectPattern(node) => {
      if let Some(rest) = &node.rest {
        identifiers = extract_identifiers(&rest.argument, identifiers);
      } else {
        for prop in &node.properties {
          identifiers = extract_identifiers(&prop.value, identifiers)
        }
      }
    }
    BindingPatternKind::ArrayPattern(node) => {
      for element in &node.elements {
        if let Some(element) = element {
          identifiers = extract_identifiers(element, identifiers);
        }
      }
    }
    BindingPatternKind::AssignmentPattern(node) => {
      identifiers = extract_identifiers(&node.left, identifiers);
    }
  }
  identifiers
}

pub fn walk_block_declarations<'a>(
  body: &'a oxc_allocator::Vec<Statement>,
  mut on_ident: impl FnMut(&'a BindingIdentifier) + 'a,
) {
  for stmt in body {
    if let Statement::VariableDeclaration(stmt) = stmt {
      if stmt.declare {
        continue;
      }
      for decl in &stmt.declarations {
        for id in extract_identifiers(&decl.id, Vec::new()) {
          on_ident(id)
        }
      }
    } else if let Statement::FunctionDeclaration(stmt) = stmt {
      if stmt.declare {
        continue;
      }
      if let Some(id) = &stmt.id {
        on_ident(id);
      }
    } else if let Statement::ClassDeclaration(stmt) = stmt {
      if stmt.declare {
        continue;
      }
      if let Some(id) = &stmt.id {
        on_ident(id);
      }
    } else if let Statement::ForStatement(stmt) = stmt {
      walk_for_statement(Either3::A(&stmt), true, &mut on_ident);
    } else if let Statement::ForInStatement(stmt) = stmt {
      walk_for_statement(Either3::B(&stmt), true, &mut on_ident);
    } else if let Statement::ForOfStatement(stmt) = stmt {
      walk_for_statement(Either3::C(&stmt), true, &mut on_ident);
    }
  }
}

pub fn walk_for_statement<'a>(
  stmt: Either3<&'a ForStatement, &'a ForInStatement, &'a ForOfStatement>,
  is_var: bool,
  on_ident: &mut impl FnMut(&'a BindingIdentifier),
) {
  let variable = if let Either3::A(stmt) = stmt
    && let Some(ForStatementInit::VariableDeclaration(stmt)) = &stmt.init
  {
    Some(stmt.as_ref())
  } else if let Either3::B(stmt) = stmt
    && let ForStatementLeft::VariableDeclaration(stmt) = &stmt.left
  {
    Some(stmt.as_ref())
  } else if let Either3::C(stmt) = stmt
    && let ForStatementLeft::VariableDeclaration(stmt) = &stmt.left
  {
    Some(stmt.as_ref())
  } else {
    None
  };
  if let Some(variable) = variable
    && if let VariableDeclarationKind::Var = variable.kind {
      is_var
    } else {
      !is_var
    }
  {
    for decl in &variable.declarations {
      for id in extract_identifiers(&decl.id, Vec::new()) {
        on_ident(id)
      }
    }
  }
}

File: packages/compiler-oxc/src/utils/walk.ts
import {
  isForStatement,
  isFunctionType,
  isIdentifier,
  isReferencedIdentifier,
} from './check'
import { extractIdentifiers } from './extract'
import { TS_NODE_TYPES } from './utils'

import type {
  BlockStatement,
  ForInStatement,
  ForOfStatement,
  ForStatement,
  Function,
  IdentifierName,
  Node,
  Program,
} from 'oxc-parser'

type WalkerContext<T = Node> = {
  skip: () => void
  remove: () => void
  replace: (node: T) => void
}

type SyncHandler<T = Node> = (
  this: WalkerContext<T>,
  node: T,
  parent: T | null,
  key: string | number | symbol | null | undefined,
  index: number | null | undefined,
) => void

export class SyncWalker<T = Node> {
  enter: SyncHandler<T>
  leave: SyncHandler<T>
  context: WalkerContext<T>
  constructor(enter: SyncHandler<T>, leave: SyncHandler<T>) {
    this.context = {
      skip: () => (this.should_skip = true),
      remove: () => (this.should_remove = true),
      replace: (node) => (this.replacement = node),
    }

    this.enter = enter

    this.leave = leave
  }

  replace(
    parent?: T | null,
    prop?: keyof T | null,
    index?: number | null,
    node?: T,
  ) {
    if (parent && prop) {
      if (index != null) {
        ;(parent[prop] as any)[index] = node!
      } else {
        ;(parent[prop] as any) = node
      }
    }
  }

  remove(parent?: T | null, prop?: keyof T | null, index?: number | null) {
    if (parent && prop) {
      if (index !== null && index !== undefined) {
        ;(parent[prop] as any).splice(index, 1)
      } else {
        delete parent[prop]
      }
    }
  }

  visit(
    node: T,
    parent: T | null,
    prop?: keyof T,
    index?: number | null,
  ): T | null {
    if (node) {
      if (this.enter) {
        const _should_skip = this.should_skip
        const _should_remove = this.should_remove
        const _replacement = this.replacement
        this.should_skip = false
        this.should_remove = false
        this.replacement = null

        this.enter.call(this.context, node, parent, prop, index)

        if (this.replacement) {
          node = this.replacement
          this.replace(parent, prop, index, node)
        }

        if (this.should_remove) {
          this.remove(parent, prop, index)
        }

        const skipped = this.should_skip
        const removed = this.should_remove

        this.should_skip = _should_skip
        this.should_remove = _should_remove
        this.replacement = _replacement

        if (skipped) return node
        if (removed) return null
      }

      let key: keyof T

      // eslint-disable-next-line no-restricted-syntax
      for (key in node) {
        const value = node[key]

        if (value && typeof value === 'object') {
          if (Array.isArray(value)) {
            const nodes = value
            for (let i = 0; i < nodes.length; i += 1) {
              const item = nodes[i]
              if (isNode(item) && !this.visit(item as any, node, key, i)) {
                // removed
                i--
              }
            }
          } else if (isNode(value)) {
            this.visit(value as any, node, key, null)
          }
        }
      }

      if (this.leave) {
        const _replacement = this.replacement
        const _should_remove = this.should_remove
        this.replacement = null
        this.should_remove = false

        this.leave.call(this.context, node, parent, prop, index)

        if (this.replacement) {
          node = this.replacement
          this.replace(parent, prop, index, node)
        }

        if (this.should_remove) {
          this.remove(parent, prop, index)
        }

        const removed = this.should_remove

        this.replacement = _replacement
        this.should_remove = _should_remove

        if (removed) return null
      }
    }

    return node
  }
  should_skip = false
  should_remove = false
  replacement: T | null = null
}

function isNode<T = Node>(value: any): value is T {
  return (
    value !== null &&
    typeof value === 'object' &&
    'type' in value &&
    typeof value.type === 'string'
  )
}

interface WalkThis<T> {
  skip: () => void
  remove: () => void
  replace: (node: T) => void
}

type WalkCallback<T, R> = (
  this: WalkThis<T>,
  node: T,
  parent: T | null | undefined,
  key: string | null | undefined,
  index: number | null | undefined,
) => R

interface WalkHandlers<T, R> {
  enter?: WalkCallback<T, R>
  leave?: WalkCallback<T, R>
}

export function walk<T>(
  ast: T,
  { enter, leave }: { enter: SyncHandler<T>; leave: SyncHandler<T> },
): T | null {
  const instance = new SyncWalker<T>(enter, leave)
  return instance.visit(ast, null)
}

/**
 * Walks the AST and applies the provided handlers.
 *
 * @template T - The type of the AST node.
 * @param {T} node - The root node of the AST.
 * @param {WalkHandlers<T, void>} hooks - The handlers to be applied during the walk.
 * @returns {T | null} - The modified AST node or null if the node is removed.
 */
export const walkAST: <T = Node>(
  node: NoInfer<T>,
  hooks: WalkHandlers<T, void>,
) => T | null = walk as any

/**
 * Modified from https://github.com/vuejs/core/blob/main/packages/compiler-core/src/babelUtils.ts
 * To support browser environments and JSX.
 *
 * https://github.com/vuejs/core/blob/main/LICENSE
 */

/**
 * Return value indicates whether the AST walked can be a constant
 */
export function walkIdentifiers(
  root: Node,
  onIdentifier: (
    node: IdentifierName,
    parent: Node | null | undefined,
    parentStack: Node[],
    isReference: boolean,
    isLocal: boolean,
  ) => void,
  includeAll = false,
  parentStack: Node[] = [],
  knownIds: Record<string, number> = Object.create(null),
): void {
  const rootExp =
    root.type === 'Program'
      ? root.body[0].type === 'ExpressionStatement' && root.body[0].expression
      : root

  walkAST<Node>(root, {
    enter(node: Node & { scopeIds?: Set<string> }, parent) {
      parent && parentStack.push(parent)
      if (
        parent &&
        parent.type.startsWith('TS') &&
        !TS_NODE_TYPES.includes(parent.type as any)
      ) {
        return this.skip()
      }
      if (isIdentifier(node)) {
        const isLocal = !!knownIds[node.name]
        const isRefed = isReferencedIdentifier(node, parent, parentStack)
        if (includeAll || (isRefed && !isLocal)) {
          onIdentifier(node, parent, parentStack, isRefed, isLocal)
        }
      } else if (node.type === 'Property' && parent?.type === 'ObjectPattern') {
        // mark property in destructure pattern
        ;(node as any).inPattern = true
      } else if (isFunctionType(node)) {
        /* v8 ignore start */
        if (node.scopeIds) {
          node.scopeIds.forEach((id) => markKnownIds(id, knownIds))
          /* v8 ignore end */
        } else {
          // walk function expressions and add its arguments to known identifiers
          // so that we don't prefix them
          walkFunctionParams(node, (id) =>
            markScopeIdentifier(node, id, knownIds),
          )
        }
      } else if (node.type === 'BlockStatement') {
        /* v8 ignore start */
        if (node.scopeIds) {
          node.scopeIds.forEach((id) => markKnownIds(id, knownIds))
          /* v8 ignore end */
        } else {
          // #3445 record block-level local variables
          walkBlockDeclarations(node, (id) =>
            markScopeIdentifier(node, id, knownIds),
          )
        }
      } else if (node.type === 'CatchClause' && node.param) {
        for (const id of extractIdentifiers(node.param)) {
          markScopeIdentifier(node, id, knownIds)
        }
      } else if (isForStatement(node)) {
        walkForStatement(node, false, (id) =>
          markScopeIdentifier(node, id, knownIds),
        )
      }
    },
    leave(node: Node & { scopeIds?: Set<string> }, parent) {
      parent && parentStack.pop()
      if (node !== rootExp && node.scopeIds) {
        for (const id of node.scopeIds) {
          knownIds[id]--
          if (knownIds[id] === 0) {
            delete knownIds[id]
          }
        }
      }
    },
  })
}

export function walkFunctionParams(
  node: Function,
  onIdent: (id: IdentifierName) => void,
): void {
  for (const p of node.params) {
    for (const id of extractIdentifiers(p)) {
      onIdent(id)
    }
  }
}

export function walkBlockDeclarations(
  block: BlockStatement | Program,
  onIdent: (node: IdentifierName) => void,
): void {
  for (const stmt of block.body) {
    if (stmt.type === 'VariableDeclaration') {
      if (stmt.declare) continue
      for (const decl of stmt.declarations) {
        for (const id of extractIdentifiers(decl.id)) {
          onIdent(id)
        }
      }
    } else if (
      stmt.type === 'FunctionDeclaration' ||
      stmt.type === 'ClassDeclaration'
    ) {
      /* v8 ignore next */
      if (stmt.declare || !stmt.id) continue
      onIdent(stmt.id)
    } else if (isForStatement(stmt)) {
      walkForStatement(stmt, true, onIdent)
    }
  }
}

function walkForStatement(
  stmt: ForStatement | ForOfStatement | ForInStatement,
  isVar: boolean,
  onIdent: (id: IdentifierName) => void,
) {
  const variable = stmt.type === 'ForStatement' ? stmt.init : stmt.left
  if (
    variable &&
    variable.type === 'VariableDeclaration' &&
    (variable.kind === 'var' ? isVar : !isVar)
  ) {
    for (const decl of variable.declarations) {
      for (const id of extractIdentifiers(decl.id)) {
        onIdent(id)
      }
    }
  }
}

function markKnownIds(name: string, knownIds: Record<string, number>) {
  if (name in knownIds) {
    knownIds[name]++
  } else {
    knownIds[name] = 1
  }
}

function markScopeIdentifier(
  node: Node & { scopeIds?: Set<string> },
  child: IdentifierName,
  knownIds: Record<string, number>,
) {
  const { name } = child
  /* v8 ignore start */
  if (node.scopeIds && node.scopeIds.has(name)) {
    return
  }
  /* v8 ignore end */
  markKnownIds(name, knownIds)
  ;(node.scopeIds || (node.scopeIds = new Set())).add(name)
}
